tokio-executor/src/park.rs

```diff
@@ -237,37 +238,56 @@ impl Inner {
     fn park(&self, timeout: Option<Duration>) -> Result<(), ParkError> {
         // If currently notified, then we skip sleeping. This is checked outside
         // of the lock to avoid acquiring a mutex if not necessary.
-        match self.state.compare_and_swap(NOTIFY, IDLE, Ordering::SeqCst) {
-            NOTIFY => return Ok(()),
-            IDLE => {},
-            _ => unreachable!(),
+        if self.state.compare_exchange(NOTIFY, IDLE, SeqCst, SeqCst).is_ok() {
+            return Ok(());
+        }
+
+        // If the duration is zero, then there is no need to actually block
+        if let Some(ref dur) = timeout {
+            if *dur == Duration::from_millis(0) {
+                return Ok(());
+            }
         }
 
         // The state is currently idle, so obtain the lock and then try to
         // transition to a sleeping state.
         let mut m = self.mutex.lock().unwrap();
 
         // Transition to sleeping
-        match self.state.compare_and_swap(IDLE, SLEEP, Ordering::SeqCst) {
-            NOTIFY => {
+        match self.state.compare_exchange(IDLE, SLEEP, SeqCst, SeqCst) {
+            Ok(_) => {}
+            Err(NOTIFY) => {
                 // Notified before we could sleep, consume the notification and
                 // exit
-                self.state.store(IDLE, Ordering::SeqCst);
+                self.state.store(IDLE, SeqCst);
                 return Ok(());
             }
-            IDLE => {},
             _ => unreachable!(),
         }
 
-        m = match timeout {
-            Some(timeout) => self.condvar.wait_timeout(m, timeout).unwrap().0,
-            None => self.condvar.wait(m).unwrap(),
+        match timeout {
+            Some(timeout) => {
+                m = self.condvar.wait_timeout(m, timeout).unwrap().0;
```

Comment:
Equivalent to: https://github.com/rust-lang/rust/blob/6e5e63a48c3be7de2faf914c2a9e194ff7cede9e/src/libstd/thread/mod.rs#L899